ONSdigital/eq-author-app
# 🛠 EAR 1227 - Validate routing destinations

### What is the context of this PR?

[Ticket](https://collaborate2.ons.gov.uk/jira/browse/EAR-1227)

Fix validation for routing destinations - make else destination
validation consistent with routing rules' destination validation

Remove 'Default' destination type and use `null` instead to signify
empty destination in need of selection.

Don't delete answer IDs from routing rules when question pages / sections deleted -
use AJV validation rules instead to check if answers exist and report
accordingly

Move out destination validation in AJV schema so it can be re-used across else +
routing rules - add `validationErrorInfo` resolver to `Destination2` type

Delete redundant validation code in `updateRoutingRule2` and `updateRouting2`
mutations - should all now be handled by AJV schema; cases previously caught & thrown exceptions
for will all cause validation errors for clients

Factor out common destination updating logic into new businessLogic file `updateDestination.js`

### How to review

- Check both `else` and routing rule destination selectors have "moved" and "deleted" validation errors appropriately when you move / delete target question pages
- Same but with a section as the target
- Check I ain’t broke other validation for routing (e.g. on expressions themselves)

### What to do after everything is green

1. - [ ] Bring this branch up-to-date with Origin/Master
2. - [ ] If there are a lot of commits or some are not helpful to read, squash them down
3. - [ ] Click the **Merge** button on this pull request
4. - [ ] Does this change mean the **Capability examples** questionnaire should be updated?
5. - [ ] Move the Jira ticket for this task into the next stage of the process
